flocker/testtools/__init__.py

```diff
@@ -868,6 +870,76 @@ def __str__(self):
         return base + " and output:\n" + lines
 
 
+class _LoggingProcessProtocol(ProcessProtocol):
+    """
+    A ``ProcessProtocol`` that stores output from the subprocess and also logs
+    output as its received as eliot messages.
+
+    Intended to be used by ``logged_run_process``.
+    """
+
+    def __init__(self, deferred):
+        """
+        Construct a ``_LoggingProcessProtocol``.
+
+        :param deferred: A ``Deferred`` that will fire with ``(reason, output)``
+            when the process ends, where ``reason`` is a ``Failure`` with the
+            reason for the process ending (see ``IProcessProtocol``), and
+            ``output`` are the bytes outputted by the process (both to stdout
+            and stderr).
+        """
+        self._deferred = deferred
+        self._output_buffer = StringIO()
+
+    def _log_output(self, data):
+        """
+        Log ``data`` as an eliot message.
+        """
+        # TODO: This should probably be within some sort of action context.
+        # TODO: Do we want to line buffer this?
+        Message.new(output=data).write()
+
+    def outReceived(self, data):
+        self._output_buffer.write(data)
+        self._log_output(data)
+
+    def errReceived(self, data):
```

Comment:
It might be useful to record whether the output comes from `stdout` or `stderr`. :question: